moban/engine_factory.py

```diff
@@ -0,0 +1,156 @@
+import os
+from collections import defaultdict
+
+import moban.utils as utils
+import moban.constants as constants
+import moban.exceptions as exceptions
+from lml.plugin import PluginManager
+from moban.extensions import LibraryManager
+
+BUILTIN_EXENSIONS = [
+    "moban.filters.repr",
+    "moban.filters.github",
+    "moban.filters.text",
+    "moban.tests.files",
+]
+
+LIBRARIES = LibraryManager()
+
+
+class EngineFactory(PluginManager):
+    def __init__(self):
+        super(EngineFactory, self).__init__(
+            constants.TEMPLATE_ENGINE_EXTENSION
+        )
+
+    def get_engine(self, template_type):
+        return self.load_me_now(template_type)
+
+    def all_types(self):
+        return list(self.registry.keys())
+
+    def raise_exception(self, key):
+        raise exceptions.NoThirdPartyEngine(key)
+
+
+ENGINES = EngineFactory()
+MOBAN_EXTENSIONS = "^moban_.+$"
+MOBAN_TEMPLATES = "^.+_mobans_pkg$"
+MOBAN_ALL = "%s|%s" % (MOBAN_EXTENSIONS, MOBAN_TEMPLATES)
+
+
+class Context(object):
+    def __init__(self, context_dirs):
+        verify_the_existence_of_directories(context_dirs)
+        self.context_dirs = context_dirs
+        self.__cached_environ_variables = dict(
+            (key, os.environ[key]) for key in os.environ
+        )
+
+    def get_data(self, file_name):
+        data = utils.open_yaml(self.context_dirs, file_name)
+        utils.merge(data, self.__cached_environ_variables)
+        return data
+
+    def get_json_data(self, file_name):
```

Comment:
json data is welcome. shall examine its file extension first? Then we have the flexibility of having yaml and json at the same time.